subscription.go

```diff
@@ -0,0 +1,219 @@
+package opcua
+
+import (
+	"context"
+	"math"
+	"time"
+
+	"github.com/gopcua/opcua/ua"
+)
+
+const (
+	DefaultSubscriptionMaxNotificationsPerPublish = 10000
+	DefaultSubscriptionLifetimeCount              = 10000
+	DefaultSubscriptionMaxKeepAliveCount          = 3000
+	DefaultSubscriptionInterval                   = 100 * time.Millisecond
+	DefaultSubscriptionPriority                   = 0
+)
+
+type Subscription struct {
+	SubscriptionID            uint32
+	RevisedPublishingInterval float64
+	RevisedLifetimeCount      uint32
+	RevisedMaxKeepAliveCount  uint32
+	Notifs                    chan *PublishNotificationData
+	c                         *Client
+}
+
+type SubscriptionParameters struct {
+	Interval                   time.Duration
+	LifetimeCount              uint32
+	MaxKeepAliveCount          uint32
+	MaxNotificationsPerPublish uint32
+	Priority                   uint8
+	Notifs                     chan *PublishNotificationData
+}
+
+func NewMonitoredItemCreateRequestWithDefaults(nodeID *ua.NodeID, attributeID ua.AttributeID, clientHandle uint32) *ua.MonitoredItemCreateRequest {
+	if attributeID == 0 {
+		attributeID = ua.AttributeIDValue
+	}
+	return &ua.MonitoredItemCreateRequest{
+		ItemToMonitor: &ua.ReadValueID{
+			NodeID:       nodeID,
+			AttributeID:  attributeID,
+			DataEncoding: &ua.QualifiedName{},
+		},
+		MonitoringMode: ua.MonitoringModeReporting,
+		RequestedParameters: &ua.MonitoringParameters{
+			ClientHandle:     clientHandle,
+			DiscardOldest:    true,
+			Filter:           nil,
+			QueueSize:        10,
+			SamplingInterval: 0.0,
+		},
+	}
+}
+
+type PublishNotificationData struct {
+	SubscriptionID uint32
+	Error          error
+	Value          interface{}
+}
+
+// Cancel() deletes the Subscription from Server and makes the Client forget it so that publishing
+// loops cannot deliver notifications to it anymore
+func (s *Subscription) Cancel() error {
+	delete(s.c.subscriptions, s.SubscriptionID)
```

Comment:
This does not look safe for concurrent use and the `Subscription` object should not modify the internal state of the `Client` directly. Use a method.

Reply:
good point